### ISSUE TITLE
matchit: match_words in ftplugin/tex.vim

### ISSUE THREAD

**Issue by Konfekt** [vim/vim#1052]
How about replacing

``` vim
if exists("loaded_matchit")
  let b:match_words .= ',\\begin\s*\({\a\+\*\=}\):\\end\s*\1'
endif " exists("loaded_matchit")
```

by

``` vim
if exists("loaded_matchit")
  setlocal matchpairs-=<:>
  let b:match_words = &l:matchpairs .
      \ ',' . '\%(^\|[ (]\)\@<=\$\$\?' . ':' . '\$\?\$\%($\|[ ).\,;\:?!\-]\)' .
      \ ',' . '\%(^\s*\)\@<=\\begin{\w\+\*\?}' . ':' . '\%(^\s*\)\@<=\\end{\w\+\*\?}'
```

For example, the regex

``` vim
\\begin\s*\({\a\+\*\=}\):\\end\s*\1
```

does not allow for indented environments such as

``` tex
  \begin{itemize}
    ...
  \end{itemize}
```

Regarding the matching of opening and closing $ ... $, this will never
be possible by a mere regex, but the above regex is a pragmatic
work-around that works if certain natural conventions are followed.

This mail was sent to Benji Fisher, the maintainer of the $VIMRUNTIME/ftplugin/tex.vim file the March 2, but yet no response.

**Comment by dkearns** [comment]
@Konfekt, are you willing to vouch for your younger self?

**Comment by Konfekt** [comment]
I've relatively recently updated this on [Stack Exchange](https://vi.stackexchange.com/a/6476)  to read

```vim
let b:match_words = get(b:, 'match_words', &l:matchpairs) .
      \ ',' . '\%(^\|[ (/]\)\zs\\(' . ':' . '\\)\ze\%($\|[ )/.\,;\:?!\-]\)' .
      \ ',' . '\%(^\|[ (/]\)\zs\\\[' . ':' . '\\\]\ze\%($\|[ )/.\,;\:?!\-]\)' .
      \ ',' . '\%(^\|[ (/]\)\zs\$[^$ ]' . ':' . '[^$ ]\zs\$\ze\%($\|[ )/.\,;\:?!\-]\)' .
      \ ',' . '\%(^\|[ (/]\)\zs\$\$' . ':' . '\$\$\ze\%($\|[ )/.\,;\:?!\-]\)' .
      \ ',' . '^\s*\zs\\begin{\([^} \t]\+\)}' . ':' . '\%(^\s*\)\zs\\end{\1}'
```

and it has been working well for all my purposes. Here's a supposedly more reliable full-fledged [plugin](https://github.com/tmhedberg/matchit/blob/master/plugin/matchit.vim) even older than this thread. I didn't yet receive a reply by @benjifisher though. I vouch for its inclusion into `ftplugin/tex.vim`. Thank you for reviving this @dkearns

**Comment by dkearns** [comment]
In what context does this value work for you?  I can't see how it can work with the distributed matchit plugin, are you using one of the alternative plugins.  `\zs` doesn't generally work in these values because of how matchit builds its internal patterns and you have to use look-behinds. 

I tried fixing it with look-arounds which only seems to work if the `.\,` in the end patterns is removed.  The `\,` causes a `E53: Unmatched \%(` error, and the `.` seems to cause the `$$` matches to fail.  I haven't tried debugging this further.

**Comment by Konfekt** [comment]
You are right; I've been using them with https://github.com/andymass/vim-matchup and thought them to be compatible; the older regex from the opening post work, though

**Comment by Konfekt** [comment]
Removing the `,` indeed makes

```vim
let b:match_words = '\%(^\|[ (/]\)\@<=\\(' . ':' . '\\)\%($\|[ )/.;\:?!\-]\)\@=' .
      \ ',' . '\%(^\|[ (/]\)\@<=\\\[' . ':' . '\\\]\%($\|[ )/.;\:?!\-]\)\@=' .
      \ ',' . '\%(^\|[ (/]\)\@<=\$[^$ ]' . ':' . '[^$ ]\@<=\$\%($\|[ )/.;\:?!\-]\)\@=' .
      \ ',' . '\%(^\|[ (/]\)\@<=\$\$' . ':' . '\$\$\%($\|[ )/.;\:?!\-]\)\@=' .
      \ ',' . '\%(^\s*\)\@<=\\begin{\(\w\+\*\?\)}' . ':' . '\%(^\s*\)\@<=\\end{\1}'
```

work; this breaks jumping between delimiters in, say `$a$, ...`, though.
Obviously `,` is used by match it to split  pairs, but how to use it in a regex if not by escaping it?

**Comment by Konfekt** [comment]
At least according to the latest official docs https://github.com/chrisbra/matchit/blob/066825be0203c061de169bdee80de8bd5d3601db/doc/matchit.txt#L307 escaping it should be way to go; curious that it doesn't bother in the opening post regex

**Comment by Konfekt** [comment]
Then again, their latest addition https://github.com/chrisbra/matchit/commit/e7270191f4d552cc4cc85cfacf07915409bcb190 allows for functions dispensing with these regex work-arounds

**Comment by dkearns** [comment]
Matching of `\(...\)` is broken.  It's falling back to using the `'matchpairs'` value of `(:)`.  I've created a [PR](https://github.com/chrisbra/matchit/pull/57) to fix this in matchit.

Escaping `,` should work and does in the general case but there's something about this example that's complicated enough to break it.  It was going to take me more than five minutes to figure out so, for the moment, I've left it alone.  I'll take a look at it soon™.

Somehow I have avoided exposure to LaTex but from what I gather `$.4 + .6$` would be valid but that fails as the opening `$` matches as an end pattern.

**Comment by Konfekt** [comment]
> I've created a https://github.com/chrisbra/matchit/pull/57 to fix this

Thank you!

> would be valid but that fails as the opening $ matches as an end pattern.

True, these are only heuristics, as `pandoc` implements them if I remember correctly, to match the ending delimiter despite the opening and ending delimiter beeing equal, by checking for a trailing space or punctuation. It's about as far as one practically gets by (unnested) regexes.

With https://github.com/chrisbra/matchit/commit/e7270191f4d552cc4cc85cfacf07915409bcb190 in place, one could copy over functions using syntax highlighting, say by https://github.com/lervag/vimtex, which work reliably inside nested delimiters such as `$ ... \text{... $ ... $ } ... $`
